client/src/app/core/ui-services/tree.service.ts

```diff
@@ -139,4 +139,60 @@ export class TreeService extends OpenSlidesComponent {
         const tree = this.makeTree(items, weightKey, parentIdKey);
         return this.traverseTree(tree);
     }
+
+    /**
+     * Reduce a list of items to nodes independent from each other in a given
+     * branch of a tree
+     *
+     * @param branch the tree to traverse
+     * @param items the items to check
+     * @returns the selection of items that belong to different branches
+     */
+    private getTopItemsFromBranch<T extends Identifiable & Displayable>(branch: OSTreeNode<T>, items: T[]): T[] {
+        const item = items.find(i => branch.item.id === i.id);
+        if (item) {
+            return [item];
+        } else if (!branch.children) {
+            return [];
+        } else {
+            return [].concat(...branch.children.map(child => this.getTopItemsFromBranch(child, items)));
+        }
+    }
+
+    /**
+     * Reduce a list of items to nodes independent from each other in a given tree
+     *
+     * @param tree the tree to traverse
+     * @param items the items to check
+     * @returns the selection of items that belong to different branches
+     */
+    public getTopItemsFromTree<T extends Identifiable & Displayable>(tree: OSTreeNode<T>[], items: T[]): T[] {
+        let results: T[] = [];
+        tree.forEach(branch => {
+            const i = this.getTopItemsFromBranch(branch, items);
+            if (i.length) {
+                results = results.concat(i);
+            }
+        });
```

Comment:
try
```ts
return tree.map(branch => {
    const top = this.getTopItemsFromBranch(branch, items);
    if (top.length) {
        return top;
    }
}) as T[];
```
(not tested)

Reply:
So, I tried copypasting your approach, and it doesn't fit. Am I to debug your approach or am I allowed to keep my oh so ugly, but working one?

Reply:
Okay. There would be `.flat` and `.flatMap` for exactly that usecase, but our setup does not yet support that.